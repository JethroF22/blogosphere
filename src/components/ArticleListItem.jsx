import React from "react";
import moment from "moment";

const ArticleListItem = props => (
  <div className="uk-card uk-card-default uk-card-hover uk-margin-medium-bottom">
    <div className="uk-card-header">
      <h3 className="article-list__title">{props.post.title}</h3>
      <p className="article-list__subtitle">
        Published on {moment(props.post.createdAt).format("MMMM Do YYYY")}
      </p>
    </div>
    <div className="uk-card-media-bottom">
      <img
        className="article-list__image"
        uk-img={`dataSrc: ${props.post.coverPhotoURL}`}
      />
    </div>
  </div>
);

export default ArticleListItem;
