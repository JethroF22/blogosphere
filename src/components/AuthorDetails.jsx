import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faUserCircle } from "@fortawesome/free-solid-svg-icons";

const AuthorDetails = props => (
  <div className="details">
    {props.photo ? (
      <div className="details__left">
        <img src={props.photo} />
      </div>
    ) : (
      <FontAwesomeIcon icon={faUserCircle} size="4x" />
    )}
    <div className="details__right">
      <p>Published by {props.username}</p>
      <button
        className="uk-button uk-button-default button"
        onClick={props.followed ? null : props.followAuthor}
        disabled={props.followed}
      >
        {props.followed ? `Following ${props.username}` : `Followed`}
      </button>
    </div>
  </div>
);

export default AuthorDetails;
