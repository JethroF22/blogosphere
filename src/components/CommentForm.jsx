import React, { Component } from "react";
import { connect } from "react-redux";

import { createComment } from "../actions/blog";

class CommentForm extends Component {
  state = {
    commentBody: ""
  };

  onCommentChange = e => {
    const commentBody = e.target.value;
    this.setState(() => ({ commentBody }));
  };

  onSubmit = e => {
    e.preventDefault();

    this.props
      .createComment(
        this.state.commentBody,
        this.props.postID,
        this.props.token
      )
      .then(slug => {
        if (this.props.actionStatus === "Action successful") {
          this.setState(() => ({
            commentBody: ""
          }));
        }
      });
  };

  render() {
    return (
      <div className="form">
        <form className="uk-form-stacked" onSubmit={this.onSubmit}>
          <label htmlFor="commentBody" className="form__label">
            Leave a comment
          </label>
          <input
            type="text"
            name="commentBody"
            value={this.state.commentBody}
            onChange={this.onCommentChange}
            placeholder="Let the author know what you think..."
            className={"uk-textarea form__textarea"}
          />
          <br />
          <button
            type="submit"
            className="uk-button uk-button-default button"
            disabled={this.state.commentBody ? false : true}
          >
            Submit comment
          </button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  token: state.auth.token,
  postID: state.blog.currentPost._id,
  actionStatus: state.status.status
});

const mapDispatchToProps = dispatch => ({
  createComment: (commentBody, postID, token) =>
    dispatch(createComment(commentBody, postID, token))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CommentForm);
