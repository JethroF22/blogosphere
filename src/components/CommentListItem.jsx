import React from "react";
import moment from "moment";

const CommentListItem = props => (
  <div>
    <h3>
      {props.comment.username} ({moment(props.comment.timestamp).format(
        "Do MMM"
      )})
    </h3>
    <p>{props.comment.body}</p>
  </div>
);

export default CommentListItem;
