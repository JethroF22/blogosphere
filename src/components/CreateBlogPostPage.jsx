import React, { Component } from "react";
import { connect } from "react-redux";
import Textarea from "react-textarea-autosize";

class CreateBlogPostPage extends Component {
  state = {
    body: "",
    title: "",
    coverPhotoURL: ""
  };

  render() {
    return (
      <div>
        <h1>Create a Blog Post</h1>
        <form>
          <label htmlFor="title">Title: </label>
          <input
            type="text"
            name="title"
            value={this.state.title}
            onChange={this.onTitleChange}
          />
          <br />
          <label htmlFor="body">Body: </label>
          <Textarea
            name="body"
            onChange={this.onBodyChange}
            value={this.state.body}
          />
          <br />
          <label htmlFor="coverPhoto">Cover Photo: </label>
          <input
            type="text"
            name="coverPhoto"
            value={this.state.coverPhotoURL}
            onChange={this.onUsernameChange}
            placeholder="Image URL"
          />
          <br />
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  createArticle: (article, token) => dispatch(createArticle(article, token))
});

export default connect(
  undefined,
  mapDispatchToProps
)(CreateBlogPostPage);
