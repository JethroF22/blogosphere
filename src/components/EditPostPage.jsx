import React, { Component } from "react";
import { connect } from "react-redux";
import Textarea from "react-textarea-autosize";
import _ from "lodash";

import { editPost } from "../actions/blog";
import { getPost } from "../actions/blog";

class EditPostPage extends Component {
  state = {
    body: "",
    title: "",
    coverPhotoURL: "",
    errors: {},
    slug: this.props.match.params.slug
  };

  componentDidMount() {
    if (!this.props.post) {
      this.props.getPost(this.state.slug).then(() => {
        const state = _.pick(this.props.post, [
          "title",
          "body",
          "coverPhotoURL"
        ]);
        this.setState(prevState => ({
          ...prevState,
          ...state
        }));
      });
    }
  }

  onBodyChange = e => {
    const body = e.target.value;
    this.setState(() => ({
      body
    }));
  };

  onURLChange = e => {
    const coverPhotoURL = e.target.value;
    this.setState(() => ({
      coverPhotoURL
    }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!(this.state.body === "" && this.state.title === "")) {
      const blogPost = _.pick(this.state, ["body", "coverPhotoURL"]);
      this.props
        .editPost(blogPost, this.props.token, this.state.slug)
        .then(slug => {
          if (this.props.actionStatus === "Action successful") {
            this.props.history.push(`/blog/view/${slug}`);
          }
        });
    } else {
      const errors = {};
      ["title", "body"].forEach(field => {
        if (this.state[field] === "") {
          errors[field] = `"${field}" cannot be blank`;
        }
        this.setState(() => ({ errors }));
      });
    }
  };

  render() {
    return (
      <div>
        <h1>Edit Blog Post</h1>
        <form onSubmit={this.onSubmit}>
          {this.props.creationError && <p>{this.props.creationError}</p>}
          <label htmlFor="title">Title: </label>
          <input
            type="text"
            name="title"
            value={this.state.title}
            disabled={true}
          />
          {this.state.errors.title && <p>{this.state.errors.title}</p>}
          <br />
          <label htmlFor="body">Body: </label>
          <Textarea
            name="body"
            onChange={this.onBodyChange}
            value={this.state.body}
          />
          {this.state.errors.body && <p>{this.state.errors.body}</p>}
          <br />
          <label htmlFor="coverPhoto">Cover Photo: </label>
          <input
            type="text"
            name="coverPhoto"
            value={this.state.coverPhotoURL}
            onChange={this.onURLChange}
            placeholder="Image URL (optional)"
          />
          <br />
          <button type="submit">Update Post</button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  actionStatus: state.status.status,
  creationError: state.error.message,
  post: state.blog.currentPost,
  token: state.auth.token
});

const mapDispatchToProps = dispatch => ({
  editPost: (article, token, slug) => dispatch(editPost(article, token, slug)),
  getPost: slug => dispatch(getPost(slug))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPostPage);
