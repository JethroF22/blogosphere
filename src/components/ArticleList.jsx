import React from "react";
import { Link } from "react-router-dom";

import ArticleListItem from "./ArticleListItem";

const ArticleList = props => (
  <div className="article-list">
    {props.posts.map(post => (
      <Link
        to={`/blog/view/${post.slug}`}
        key={Math.floor(Math.random() * 99999)}
      >
        <ArticleListItem post={post} />
      </Link>
    ))}
  </div>
);

export default ArticleList;
