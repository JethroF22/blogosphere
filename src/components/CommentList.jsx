import React from "react";
import CommentListItem from "./CommentListItem";

const CommentList = props => (
  <div className="comments">
    {props.comments.map(comment => <CommentListItem comment={comment} />)}
  </div>
);

export default CommentList;
