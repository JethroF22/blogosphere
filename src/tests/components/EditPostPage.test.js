import React from "react";
import { shallow } from "enzyme";

import { EditPostPage } from "../../components/EditPostPage";

let wrapper, getPost, slug;

beforeEach(() => {
  slug = "slug";
  const match = {
    params: {
      slug
    }
  };
  getPost = jest.fn().mockResolvedValueOnce({});
  wrapper = shallow(<EditPostPage getPost={getPost} match={match} />);
});

test("should render component correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should fetch post if no post is passed as prop", () => {
  expect(getPost).toHaveBeenLastCalledWith(slug);
});
