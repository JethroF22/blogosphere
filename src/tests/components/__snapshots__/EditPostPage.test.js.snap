// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly 1`] = `
<div>
  <h1>
    Edit Blog Post
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="title"
    >
      Title: 
    </label>
    <input
      disabled={true}
      name="title"
      type="text"
      value=""
    />
    <br />
    <label
      htmlFor="body"
    >
      Body: 
    </label>
    <TextareaAutosize
      inputRef={[Function]}
      name="body"
      onChange={[Function]}
      onHeightChange={[Function]}
      useCacheForDOMMeasurements={false}
      value=""
    />
    <br />
    <label
      htmlFor="coverPhoto"
    >
      Cover Photo: 
    </label>
    <input
      name="coverPhoto"
      onChange={[Function]}
      placeholder="Image URL (optional)"
      type="text"
      value=""
    />
    <br />
    <button
      type="submit"
    >
      Update Post
    </button>
  </form>
</div>
`;
